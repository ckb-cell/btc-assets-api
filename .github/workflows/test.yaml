---
name: Unit Tests

on:
  pull_request:
    branches:
      - main
      - develop

jobs:
  test:
    runs-on: ubuntu-latest

    services:
      redis:
        image: redis
        options: >-
          --health-cmd "redis-cli ping"
          --health-interval 10s
          --health-timeout 5s
          --health-retries 5
        ports:
          - 6379:6379

    steps:
      - name: Checkout Code
        uses: actions/checkout@v3

      - name: Install PNPM
        uses: pnpm/action-setup@v3
        with:
          version: 8

      - name: Use Node.js
        uses: actions/setup-node@v3
        with:
          node-version: "21"
          cache: "pnpm"

      - name: Install dependencies
        run: pnpm install

      - name: Run Unit Tests
        env:
          JWT_SECRET: ${{ secrets.JWT_SECRET }}
          BITCOIN_JSON_RPC_URL: ${{ secrets.BITCOIN_JSON_RPC_URL }}
          BITCOIN_JSON_RPC_USERNAME: ${{ secrets.BITCOIN_JSON_RPC_USERNAME }}
          BITCOIN_JSON_RPC_PASSWORD: ${{ secrets.BITCOIN_JSON_RPC_PASSWORD }}
          BITCOIN_ELECTRS_API_URL: ${{ secrets.BITCOIN_ELECTRS_API_URL }}
          BITCOIN_SPV_SERVICE_URL: ${{ secrets.BITCOIN_SPV_SERVICE_URL }}
          CKB_RPC_URL: ${{ secrets.CKB_RPC_URL }}
          CKB_INDEXER_URL: ${{ secrets.CKB_INDEXER_URL }}
          PAYMASTER_PRIVATE_KEY: ${{ secrets.PAYMASTER_PRIVATE_KEY }}
          REDIS_URL: redis://localhost:6379
          CI_REDIS_URL: redis://localhost:6379
        run: pnpm test
