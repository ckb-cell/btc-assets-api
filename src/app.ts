import { FastifyInstance } from 'fastify';
import { AxiosError } from 'axios';
import bitcoinRoutes from './routes/bitcoin';
import { env } from './env';
import tokenRoutes from './routes/token';
import swaggerPlugin from './plugins/swagger';
import jwtPlugin from './plugins/jwt';
import cachePlugin from './plugins/cache';
import rateLimitPlugin from './plugins/rate-limit';
import * as Sentry from '@sentry/node';

if (env.SENTRY_DSN_URL && env.NODE_ENV !== 'development') {
  Sentry.init({
    dsn: env.SENTRY_DSN_URL,
    tracesSampleRate: 1.0,
  });
}

async function routes(fastify: FastifyInstance) {
  fastify.register(rateLimitPlugin);
  fastify.register(swaggerPlugin);
  fastify.register(jwtPlugin);
  fastify.register(cachePlugin);

  fastify.register(tokenRoutes, { prefix: '/token' });
  fastify.register(bitcoinRoutes, { prefix: '/bitcoin/v1' });

  fastify.setErrorHandler((error, _, reply) => {
    Sentry.captureException(error);
    if (error instanceof AxiosError) {
      reply
        .status(error.response?.status || 500)
        .send({ ok: false, error: error.response?.data.error ?? error.message });
      return;
    }
    reply
      .status(error.statusCode ?? 500)
      .send({ ok: false, statusCode: error.statusCode, message: error.message });
  });
}

export default routes;
