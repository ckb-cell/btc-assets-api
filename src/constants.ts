export const CUSTOM_HEADERS = {
  ApiCache: 'x-api-cache',
  ResponseCacheable: 'x-response-cacheable',
};

export enum ApiCacheStatus {
  Hit = 'HIT',
  Miss = 'MISS',
}

export const JWT_IGNORE_URLS = ['/token', '/docs'];
