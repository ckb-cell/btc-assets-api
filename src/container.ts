import { createContainer, InjectionMode, asValue, asClass, asFunction } from 'awilix';
import { Redis } from 'ioredis';
import pino from 'pino';
import Bitcoind from './services/bitcoind';
import ElectrsAPI from './services/electrs';
import { env } from './env';
import TransactionManager from './services/transaction';
import Paymaster from './services/paymaster';
import { RPC as CkbRPC, Indexer as CkbIndexer } from '@ckb-lumos/lumos';
import Unlocker from './services/unlocker';

export interface Cradle {
  env: typeof env;
  logger: pino.BaseLogger;
  redis: Redis;
  ckbRpc: CkbRPC;
  ckbIndexer: CkbIndexer;
  bitcoind: Bitcoind;
  electrs: ElectrsAPI;
  paymaster: Paymaster;
  unlocker: Unlocker;
  transactionManager: TransactionManager;
}

const container = createContainer<Cradle>({
  injectionMode: InjectionMode.PROXY,
  strict: true,
});

container.register({
  env: asValue(env),
  logger: asValue(pino()),
  redis: asValue(
    new Redis(env.REDIS_URL, {
      maxRetriesPerRequest: null,
    }),
  ),
  ckbRpc: asFunction(() => new CkbRPC(env.CKB_RPC_URL)).singleton(),
  ckbIndexer: asFunction(() => new CkbIndexer(env.CKB_INDEXER_URL)).singleton(),
  bitcoind: asClass(Bitcoind).singleton(),
  electrs: asClass(ElectrsAPI).singleton(),
  paymaster: asClass(Paymaster).singleton(),
  transactionManager: asClass(TransactionManager).singleton(),
  unlocker: asClass(Unlocker).singleton(),
});

export default container;
