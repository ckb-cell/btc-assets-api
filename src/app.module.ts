import { Module } from '@nestjs/common';
import { AppService } from './app.service';
import { TokenModule } from './modules/token/token.module';
import { BitcoinModule } from './modules/bitcoin/bitcoin.module';
import { ConfigModule, ConfigService } from '@nestjs/config';
import { envSchema } from './env';
import { SentryModule } from '@ntegral/nestjs-sentry';
import { nodeProfilingIntegration } from '@sentry/profiling-node';
import { CacheModule } from '@nestjs/cache-manager';
import type { RedisClientOptions } from 'redis';
import * as redisStore from 'cache-manager-redis-store';
import { StoreConfig } from 'cache-manager';

@Module({
  imports: [
    ConfigModule.forRoot({
      isGlobal: true,
      validate: envSchema.parse,
    }),
    CacheModule.registerAsync<RedisClientOptions>({
      isGlobal: true,
      imports: [ConfigModule],
      inject: [ConfigService],
      useFactory: async (configService: ConfigService) =>
        ({
          store: redisStore,
          url: configService.get('REDIS_URL'),
          ttl: 10,
        }) as StoreConfig,
    }),
    SentryModule.forRootAsync({
      imports: [ConfigModule],
      inject: [ConfigService],
      useFactory: async (configService: ConfigService) => ({
        dsn: configService.get('SENTRY_DSN'),
        environment: configService.get('NODE_ENV'),
        tracesSampleRate: 0.5,
        profilesSampleRate: 0.5,
        integrations: [nodeProfilingIntegration()],
        logLevels:
          configService.get('NODE_ENV') === 'production'
            ? ['warn', 'error']
            : ['debug', 'log', 'warn', 'error'],
        beforeSend: (event) => {
          if (event.level === 'error') {
            return event;
          }
          return null;
        },
      }),
    }),
    TokenModule,
    BitcoinModule,
  ],
  controllers: [],
  providers: [AppService],
})
export class AppModule {}
