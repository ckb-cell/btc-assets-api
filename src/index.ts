import process from 'node:process';
import { env } from './env';
import { buildFastify } from './app';

const port = parseInt(env.PORT || '3000', 10);
const host = env.ADDRESS || '0.0.0.0';

const app = buildFastify();

app.listen({ port, host }, (err, address) => {
  if (err) {
    console.error(err);
    process.exit(1);
  }

  // eslint-disable-next-line no-console
  console.log(`Server listening at ${address}`);
});
