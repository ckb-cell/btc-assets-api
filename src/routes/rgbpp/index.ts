import { FastifyPluginCallback } from 'fastify';
import { Server } from 'http';
import container from '../../container';
import transactionRoutes from './transaction';
import { ZodTypeProvider } from 'fastify-type-provider-zod';

const rgbppRoutes: FastifyPluginCallback<Record<never, never>, Server, ZodTypeProvider> = (fastify, _, done) => {
  fastify.decorate('transactionManager', container.resolve('transactionManager'));
  fastify.register(transactionRoutes, { prefix: '/transaction' });
  done();
};

export default rgbppRoutes;
