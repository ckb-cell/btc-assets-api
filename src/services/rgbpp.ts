import { UTXO } from './bitcoin/schema';
import pLimit from 'p-limit';
import asyncRetry from 'async-retry';
import { Cradle } from '../container';
import { CKBIndexerQueryOptions } from '@ckb-lumos/ckb-indexer/lib/type';
import { buildRgbppLockArgs, genRgbppLockScript } from '@rgbpp-sdk/ckb';
import * as Sentry from '@sentry/node';
import { Script } from '@ckb-lumos/lumos';
import { Job } from 'bullmq';
import { z } from 'zod';
import { Cell } from '../routes/rgbpp/types';
import BaseQueueWorker from './base/queue-worker';
import DataCache from './base/data-cache';

type RgbppUtxoCellsPair = {
  utxo: UTXO;
  cells: Cell[];
};

interface IRgbppCollectRequest {
  btcAddress: string;
  utxos: UTXO[];
}

interface IRgbppCollectJobReturn {
  [key: string]: Cell[];
}

export interface IProcessCallbacks {
  onActive?: (job: Job<IRgbppCollectRequest>) => void;
  onCompleted?: (job: Job<IRgbppCollectRequest>) => void;
  onFailed?: (job: Job<IRgbppCollectRequest> | undefined, err: Error) => void;
}

export const RGBPP_COLLECTOR_QUEUE_NAME = 'rgbpp-collector-queue';

export default class RgbppCollector extends BaseQueueWorker<IRgbppCollectRequest, IRgbppCollectJobReturn> {
  private limit: pLimit.Limit;
  private dataCahe: DataCache<IRgbppCollectJobReturn>;

  constructor(private cradle: Cradle) {
    super({
      name: RGBPP_COLLECTOR_QUEUE_NAME,
      connection: cradle.redis,
      worker: {
        lockDuration: 60_000,
        removeOnComplete: { count: 0 },
        removeOnFail: { count: 0 },
      },
    });
    this.dataCahe = new DataCache(cradle.redis, {
      prefix: 'rgbpp-collector-data',
      schema: z.record(z.array(Cell)),
    });
    this.limit = pLimit(cradle.env.CKB_RPC_MAX_CONCURRENCY);
  }

  /**
   * Get the rgbpp cells from cache
   * @param btcAddress - the btc address
   */
  public async getRgbppCellsFromCache(btcAddress: string) {
    const data = await this.dataCahe.get(btcAddress);
    if (!data) {
      return null;
    }
    return Object.values(data).flat();
  }

  /**
   * Get the cells for the utxo, use ckb indexer to query the cells
   * @param utxo - the utxo to collect
   * @param typeScript - the type script to filter the cells
   */
  private async getRgbppCellsByUtxo(utxo: UTXO, typeScript?: Script): Promise<Cell[]> {
    try {
      const { txid, vout } = utxo;
      const args = buildRgbppLockArgs(vout, txid);

      const query: CKBIndexerQueryOptions = {
        lock: genRgbppLockScript(args, process.env.NETWORK === 'mainnet'),
      };

      if (typeScript) {
        query.type = typeScript;
      }

      const collector = this.cradle.ckb.indexer.collector(query).collect();
      const cells: Cell[] = [];
      for await (const cell of collector) {
        cells.push(cell);
      }
      return cells;
    } catch (e) {
      Sentry.withScope((scope) => {
        scope.captureException(e);
      });
      throw e;
    }
  }

  /**
   * Collect the cells for the utxos, return the utxo and the cells
   * @param utxos - the utxos to collect
   * @param typeScript - the type script to filter the cells
   */
  public async collectRgbppUtxoCellsPairs(utxos: UTXO[], typeScript?: Script): Promise<RgbppUtxoCellsPair[]> {
    const cells = await Promise.all(
      utxos.map((utxo) => {
        return this.limit(() =>
          asyncRetry(
            async () => {
              const cells = await this.getRgbppCellsByUtxo(utxo, typeScript);
              return { utxo, cells };
            },
            { retries: 2 },
          ),
        );
      }),
    );

    return cells.filter(({ cells }) => cells.length > 0);
  }

  /**
   * Enqueue a collect job to the queue
   * @param utxos - the utxos to collect
   */
  public async enqueueCollectJob(
    btcAddress: string,
    utxos: UTXO[],
    allowDuplicate?: boolean,
  ): Promise<Job<IRgbppCollectRequest>> {
    let jobId = btcAddress;
    if (allowDuplicate) {
      jobId = `${btcAddress}:${Date.now()}`;
    }
    return this.addJob(jobId, { btcAddress, utxos });
  }

  /**
   * Process the collect job, collect the cells for the utxos
   * concurrently controlled by the CKB_RPC_MAX_CONCURRENCY
   * retry 2 times if failed, and return the utxo and cells
   */
  public async process(job: Job<IRgbppCollectRequest>) {
    const { btcAddress, utxos } = job.data;
    const pairs = await this.collectRgbppUtxoCellsPairs(utxos);
    const data = pairs.reduce((acc, { utxo, cells }) => {
      const key = `${utxo.txid}:${utxo.vout}`;
      acc[key] = cells;
      return acc;
    }, {} as IRgbppCollectJobReturn);
    this.dataCahe.set(btcAddress, data);
    return data;
  }
}
