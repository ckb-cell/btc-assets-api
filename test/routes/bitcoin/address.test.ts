import { describe, expect, test, beforeEach, vi } from 'vitest';
import { buildFastify } from '../../../src/app';
import { afterEach } from 'node:test';
import mockUTXOs from '../../__fixtures__/utxo.mock.json';
import mockRgbppUtxoPairs from '../../__fixtures__/rgbpp-utxo-pairs.mock.json';
import { RgbppUtxoCellsPair } from '../../../src/services/rgbpp';

let token: string;

describe('/bitcoin/v1/address', () => {
  beforeEach(async () => {
    const fastify = buildFastify();
    await fastify.ready();

    const response = await fastify.inject({
      method: 'POST',
      url: '/token/generate',
      payload: {
        app: 'test',
        domain: 'test.com',
      },
    });
    const data = response.json();
    token = data.token;

    await fastify.close();
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  test('Get address balance', async () => {
    const fastify = buildFastify();
    await fastify.ready();

    const utxoSyncer = fastify.container.resolve('utxoSyncer');
    vi.spyOn(utxoSyncer, 'getUtxosByAddress').mockResolvedValueOnce(mockUTXOs);
    const rgbppCollector = fastify.container.resolve('rgbppCollector');
    vi.spyOn(rgbppCollector, 'getRgbppUtxoCellsPairs').mockResolvedValueOnce(
      mockRgbppUtxoPairs as RgbppUtxoCellsPair[],
    );

    const response = await fastify.inject({
      method: 'GET',
      url: '/bitcoin/v1/address/tb1pusqtndhyt8mcrggngd0lmah4xshc0aa8vpfsttvtzxd0hxhkrwns7977yu/balance',
      headers: {
        Authorization: `Bearer ${token}`,
        Origin: 'https://test.com',
      },
    });
    const data = response.json();

    expect(response.statusCode).toBe(200);
    expect(data).toMatchSnapshot();

    await fastify.close();
  });

  test('Get address balance with min_satoshi param', async () => {
    const fastify = buildFastify();
    await fastify.ready();

    const utxoSyncer = fastify.container.resolve('utxoSyncer');
    vi.spyOn(utxoSyncer, 'getUtxosByAddress').mockResolvedValueOnce(mockUTXOs);
    const rgbppCollector = fastify.container.resolve('rgbppCollector');
    vi.spyOn(rgbppCollector, 'getRgbppUtxoCellsPairs').mockResolvedValueOnce(
      mockRgbppUtxoPairs as RgbppUtxoCellsPair[],
    );

    const response = await fastify.inject({
      method: 'GET',
      url: '/bitcoin/v1/address/tb1pusqtndhyt8mcrggngd0lmah4xshc0aa8vpfsttvtzxd0hxhkrwns7977yu/balance?min_satoshi=10000',
      headers: {
        Authorization: `Bearer ${token}`,
        Origin: 'https://test.com',
      },
    });
    const data = response.json();

    expect(response.statusCode).toBe(200);
    expect(data).toMatchSnapshot();

    await fastify.close();
  });

  test('Get address balance with invalid address', async () => {
    const fastify = buildFastify();
    await fastify.ready();

    const response = await fastify.inject({
      method: 'GET',
      url: '/bitcoin/v1/address/tb1qlrg2mhyxrq7ns5rpa6qvrvttr9674n6z0try/balance',
      headers: {
        Authorization: `Bearer ${token}`,
        Origin: 'https://test.com',
      },
    });
    const data = response.json();

    expect(response.statusCode).toBe(400);
    expect(data.message).toBe('Invalid bitcoin address');

    await fastify.close();
  });

  test('Get address unspent transaction outputs', async () => {
    const fastify = buildFastify();
    await fastify.ready();

    const utxoSyncer = fastify.container.resolve('utxoSyncer');
    vi.spyOn(utxoSyncer, 'getUtxosByAddress').mockResolvedValueOnce(mockUTXOs);
    const rgbppCollector = fastify.container.resolve('rgbppCollector');
    vi.spyOn(rgbppCollector, 'getRgbppUtxoCellsPairs').mockResolvedValueOnce(
      mockRgbppUtxoPairs as RgbppUtxoCellsPair[],
    );

    const response = await fastify.inject({
      method: 'GET',
      url: '/bitcoin/v1/address/tb1pusqtndhyt8mcrggngd0lmah4xshc0aa8vpfsttvtzxd0hxhkrwns7977yu/unspent',
      headers: {
        Authorization: `Bearer ${token}`,
        Origin: 'https://test.com',
      },
    });
    const data = response.json();

    expect(response.statusCode).toBe(200);
    expect(data).toMatchObject(mockUTXOs.filter((utxo) => utxo.status.confirmed));

    await fastify.close();
  });

  test('Get address unspent transaction outputs with only_confirmed = true', async () => {
    const fastify = buildFastify();
    await fastify.ready();

    const utxoSyncer = fastify.container.resolve('utxoSyncer');
    vi.spyOn(utxoSyncer, 'getUtxosByAddress').mockResolvedValueOnce(mockUTXOs);
    const rgbppCollector = fastify.container.resolve('rgbppCollector');
    vi.spyOn(rgbppCollector, 'getRgbppUtxoCellsPairs').mockResolvedValueOnce(
      mockRgbppUtxoPairs as RgbppUtxoCellsPair[],
    );

    const response = await fastify.inject({
      method: 'GET',
      url: '/bitcoin/v1/address/tb1pusqtndhyt8mcrggngd0lmah4xshc0aa8vpfsttvtzxd0hxhkrwns7977yu/unspent?only_confirmed=true',
      headers: {
        Authorization: `Bearer ${token}`,
        Origin: 'https://test.com',
      },
    });
    const data = response.json();

    expect(response.statusCode).toBe(200);
    expect(data).toMatchObject(mockUTXOs.filter((utxo) => utxo.status.confirmed));

    await fastify.close();
  });

  test('Get address unspent transaction outputs with only_confirmed = false', async () => {
    const fastify = buildFastify();
    await fastify.ready();

    const utxoSyncer = fastify.container.resolve('utxoSyncer');
    vi.spyOn(utxoSyncer, 'getUtxosByAddress').mockResolvedValueOnce(mockUTXOs);
    const rgbppCollector = fastify.container.resolve('rgbppCollector');
    vi.spyOn(rgbppCollector, 'getRgbppUtxoCellsPairs').mockResolvedValueOnce(
      mockRgbppUtxoPairs as RgbppUtxoCellsPair[],
    );

    const response = await fastify.inject({
      method: 'GET',
      url: '/bitcoin/v1/address/tb1pusqtndhyt8mcrggngd0lmah4xshc0aa8vpfsttvtzxd0hxhkrwns7977yu/unspent?only_confirmed=false',
      headers: {
        Authorization: `Bearer ${token}`,
        Origin: 'https://test.com',
      },
    });
    const data = response.json();

    expect(response.statusCode).toBe(200);
    expect(data).toMatchObject(mockUTXOs);

    await fastify.close();
  });

  test('Get address unspent transaction outputs with min_satoshi = 10000', async () => {
    const fastify = buildFastify();
    await fastify.ready();

    const utxoSyncer = fastify.container.resolve('utxoSyncer');
    vi.spyOn(utxoSyncer, 'getUtxosByAddress').mockResolvedValueOnce(mockUTXOs);
    const rgbppCollector = fastify.container.resolve('rgbppCollector');
    vi.spyOn(rgbppCollector, 'getRgbppUtxoCellsPairs').mockResolvedValueOnce(
      mockRgbppUtxoPairs as RgbppUtxoCellsPair[],
    );

    const response = await fastify.inject({
      method: 'GET',
      url: '/bitcoin/v1/address/tb1pusqtndhyt8mcrggngd0lmah4xshc0aa8vpfsttvtzxd0hxhkrwns7977yu/unspent?min_satoshi=10000',
      headers: {
        Authorization: `Bearer ${token}`,
        Origin: 'https://test.com',
      },
    });
    const data = response.json();

    expect(response.statusCode).toBe(200);
    expect(data).toMatchObject(mockUTXOs.filter((utxo) => utxo.value >= 10000));

    await fastify.close();
  });

  test('Get address unspent transaction outputs with onlyNonRgbppUtxos = true', async () => {
    const fastify = buildFastify();
    await fastify.ready();

    const utxoSyncer = fastify.container.resolve('utxoSyncer');
    vi.spyOn(utxoSyncer, 'getUtxosByAddress').mockResolvedValueOnce(mockUTXOs);
    const rgbppCollector = fastify.container.resolve('rgbppCollector');
    vi.spyOn(rgbppCollector, 'getRgbppUtxoCellsPairs').mockResolvedValueOnce(
      mockRgbppUtxoPairs as RgbppUtxoCellsPair[],
    );

    const response = await fastify.inject({
      method: 'GET',
      url: '/bitcoin/v1/address/tb1pusqtndhyt8mcrggngd0lmah4xshc0aa8vpfsttvtzxd0hxhkrwns7977yu/unspent?only_non_rgbpp_utxos=true',
      headers: {
        Authorization: `Bearer ${token}`,
        Origin: 'https://test.com',
      },
    });
    const data = response.json();

    expect(response.statusCode).toBe(200);
    expect(data).toMatchObject(
      mockUTXOs.filter((utxo) => {
        if (utxo.status.confirmed === false) return false;
        const pair = mockRgbppUtxoPairs.find((pair) => pair.utxo.txid === utxo.txid && pair.utxo.vout === utxo.vout);
        return !pair;
      }),
    );

    await fastify.close();
  });
});
