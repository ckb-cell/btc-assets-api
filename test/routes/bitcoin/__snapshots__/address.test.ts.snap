// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`/bitcoin/v1/address > Get address balance 1`] = `
{
  "address": "tb1pusqtndhyt8mcrggngd0lmah4xshc0aa8vpfsttvtzxd0hxhkrwns7977yu",
  "available_satoshi": 90738317,
  "dust_satoshi": 0,
  "pending_satoshi": 1426,
  "rgbpp_satoshi": 6552,
  "satoshi": 90738317,
  "total_satoshi": 90746295,
  "utxo_count": 17,
}
`;

exports[`/bitcoin/v1/address > Get address balance with min_satoshi param 1`] = `
{
  "address": "tb1pusqtndhyt8mcrggngd0lmah4xshc0aa8vpfsttvtzxd0hxhkrwns7977yu",
  "available_satoshi": 90737437,
  "dust_satoshi": 7432,
  "pending_satoshi": 1426,
  "rgbpp_satoshi": 6552,
  "satoshi": 90737437,
  "total_satoshi": 90746295,
  "utxo_count": 17,
}
`;
